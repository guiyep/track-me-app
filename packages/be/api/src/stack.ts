import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { LambdaRestApi } from 'aws-cdk-lib/aws-apigateway';
import { NAMES } from './config';
import * as path from 'path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export interface ApiProps {
  dependencies: {
    gpsSqs: string;
  };
}

export class Api extends Construct {
  public lambdaApi: NodejsFunction;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.lambdaApi = new NodejsFunction(this, NAMES.Lambda, {
      functionName: NAMES.Lambda,
      runtime: Runtime.NODEJS_LATEST,
      entry: path.join(__dirname, '../src/lambda/index.ts'),
      handler: 'handler',
    });

    new LambdaRestApi(this, NAMES.ApiGateway, {
      handler: this.lambdaApi,
      restApiName: NAMES.ApiGateway,
    });
  }
}
