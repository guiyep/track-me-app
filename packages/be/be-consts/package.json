{
  "name": "@track-me-app/be-consts",
  "version": "0.1.0",
  "main": "src/index.ts",
  "devDependencies": {
    "jest": "^29.6.1",
    "@types/jest": "^29.5.3",
    "typescript": "5.1.6"
  },
  "dependencies": {}
}
