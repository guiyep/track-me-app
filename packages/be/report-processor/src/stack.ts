import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NAMES } from './config';
import * as path from 'path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export class ReportProcessor extends Construct {
  public lambda: NodejsFunction;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.lambda = new NodejsFunction(this, NAMES.Lambda, {
      functionName: NAMES.Lambda,
      runtime: Runtime.NODEJS_LATEST,
      entry: path.join(__dirname, './lambda/index.ts'),
      handler: 'handler',
    });
  }
}
