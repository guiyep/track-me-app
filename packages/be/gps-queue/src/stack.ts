import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import { CONFIGURATIONS } from './config';
import { getConstants } from '@track-me-app/be-consts';
import type { AccessProps } from '@track-me-app/aws';
import { IFunction } from 'aws-cdk-lib/aws-lambda';

const Consts = getConstants();

export class GpsQueue extends Construct {
  public readonly queue: sqs.Queue;
  public readonly dlq: sqs.Queue;

  constructor(
    scope: Construct,
    id: string,
    props?: AccessProps & {
      listeners?: IFunction[];
    },
  ) {
    super(scope, id);

    this.dlq = new sqs.Queue(this, Consts.GpsLocationsQueue.DLQ, {
      visibilityTimeout: cdk.Duration.seconds(CONFIGURATIONS.VisibilityTimeout),
    });

    this.queue = new sqs.Queue(this, Consts.GpsLocationsQueue.QUEUE_NAME, {
      visibilityTimeout: cdk.Duration.seconds(CONFIGURATIONS.VisibilityTimeout),
      queueName: Consts.GpsLocationsQueue.QUEUE_NAME,
      deadLetterQueue: {
        queue: this.dlq,
        maxReceiveCount: 3,
      },
    });

    props?.read?.forEach((grantable) => {
      this.queue.grantConsumeMessages(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.queue.grantSendMessages(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.queue.grantConsumeMessages(grantable);
      this.queue.grantSendMessages(grantable);
    });

    const lambdaEventSource = new lambdaEventSources.SqsEventSource(
      this.queue,
      {
        batchSize: 5,
      },
    );

    props?.listeners?.forEach((listener) => {
      this.queue.grantConsumeMessages(listener);
      this.queue.grantSendMessages(listener);
      listener.addEventSource(lambdaEventSource);
    });
  }
}
