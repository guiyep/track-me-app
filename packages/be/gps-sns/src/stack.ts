import { Construct } from 'constructs';
import type * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { getConstants } from '@track-me-app/be-consts';
import { CfnOutput } from 'aws-cdk-lib';
const Consts = getConstants();

export class GpsSns extends Construct {
  constructor(
    scope: Construct,
    id: string,
    props?: {
      subscriptions?: sqs.IQueue[];
    },
  ) {
    super(scope, id);

    const topic = new sns.Topic(this, Consts.GpsSns.TOPIC_NAME);
    if (props?.subscriptions) {
      props.subscriptions.forEach((queue: sqs.IQueue) => {
        topic.addSubscription(new subscriptions.SqsSubscription(queue));
      });
    }

    new CfnOutput(this, `GpsSnsTopicArn`, {
      value: topic.topicArn,
    });
  }
}
