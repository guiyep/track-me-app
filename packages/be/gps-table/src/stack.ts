import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { getConstants } from '@track-me-app/be-consts';
import type { AccessProps } from '@track-me-app/aws';

const Consts = getConstants();

/*
schema {
  partitionKey: '[email]',
  sortKey: '[sessionId]',
  data: {
    email: string,
    displayName: string,
    avatarSrc: string,
    sessionId: string,
    lat: number,
    long:number,
    created: number, (timestamp)
    lastUpdated: number (timestamp)
  }
}
 */

export class GpsTable extends Construct {
  private readonly table: dynamodb.TableV2;

  constructor(scope: Construct, id: string, props?: AccessProps) {
    super(scope, id);

    this.table = new dynamodb.TableV2(this, Consts.GpsTable.TABLE_NAME, {
      partitionKey: {
        name: 'partitionKey',
        type: dynamodb.AttributeType.STRING,
      },
      sortKey: { name: 'sortKey', type: dynamodb.AttributeType.STRING },
      tableName: Consts.GpsTable.TABLE_NAME,
    });

    props?.read?.forEach((grantable) => {
      this.table.grantReadData(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.table.grantWriteData(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.table.grantReadWriteData(grantable);
    });
  }
}
