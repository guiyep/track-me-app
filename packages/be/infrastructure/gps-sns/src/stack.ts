import { Construct } from 'constructs';
import type * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { getConstants } from '@track-me-app/be-consts';
import { CfnOutput } from 'aws-cdk-lib';
import type { AccessProps } from '@track-me-app/aws';
const Consts = getConstants();

// filterPolicy: {
//       eventType: sns.SubscriptionFilter.stringFilter({
//         allowlist: ['UserSignedUp'],
//       }),
//     },

type Messages = typeof Consts.GpsSns.MESSAGES;

export class GpsSns extends Construct {
  private readonly topic: sns.Topic;

  constructor(
    scope: Construct,
    id: string,
    props?: AccessProps & {
      queueSubscriptions?: {
        queue: sqs.IQueue;
        listenTo?: Messages[keyof Messages][];
      }[];
    },
  ) {
    super(scope, id);

    this.topic = new sns.Topic(this, Consts.GpsSns.TOPIC_NAME);

    props?.queueSubscriptions?.forEach(({ queue, listenTo }) => {
      this.topic.addSubscription(
        new subscriptions.SqsSubscription(queue, {
          filterPolicy: {
            eventType: sns.SubscriptionFilter.stringFilter({
              allowlist: listenTo,
            }),
          },
        }),
      );
      new CfnOutput(this, `GpsSnsTopicSubscribers`, {
        value: queue.queueArn + ' - ' + JSON.stringify(listenTo),
      });
    });

    props?.read?.forEach((grantable) => {
      this.topic.grantSubscribe(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.topic.grantPublish(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.topic.grantSubscribe(grantable);
      this.topic.grantPublish(grantable);
    });

    new CfnOutput(this, `GpsSnsTopicArn`, {
      value: this.topic.topicArn,
      description: 'The ARN of the GPS SNS topic',
    });
  }
}
