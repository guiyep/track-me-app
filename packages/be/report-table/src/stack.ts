import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { getConstants } from '@track-me-app/be-consts';
import type { AccessProps } from '@track-me-app/aws';

const Consts = getConstants();

export class ReportTable extends Construct {
  private readonly table: dynamodb.TableV2;

  constructor(scope: Construct, id: string, props?: AccessProps) {
    super(scope, id);

    this.table = new dynamodb.TableV2(this, Consts.ReportTable.TABLE_NAME, {
      partitionKey: {
        name: 'partitionKey',
        type: dynamodb.AttributeType.STRING,
      },
      sortKey: { name: 'sortKey', type: dynamodb.AttributeType.STRING },
      tableName: Consts.ReportTable.TABLE_NAME,
    });

    props?.read?.forEach((grantable) => {
      this.table.grantReadData(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.table.grantWriteData(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.table.grantReadWriteData(grantable);
    });
  }
}
