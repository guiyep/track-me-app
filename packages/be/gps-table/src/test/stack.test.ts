import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GpsTable } from '../stack';
import { getConstants } from '@track-me-app/be-consts';

const Consts = getConstants();

describe('GpsTable Stack', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');
  new GpsTable(stack, 'TestGpsTable');

  test('snapshot', () => {
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('should create a DynamoDB table with correct properties', () => {
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::DynamoDB::GlobalTable', {
      TableName: Consts.GpsTable.TABLE_NAME,
    });
  });
});
