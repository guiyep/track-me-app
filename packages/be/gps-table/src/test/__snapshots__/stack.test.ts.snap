// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GpsTable Stack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestGpsTableGpsLocationsC47CFCB6": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "partitionKey",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sortKey",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "partitionKey",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sortKey",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
        "TableName": "GpsLocations",
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
